lib/phoenix_live_view/html_engine.ex

```diff
@@ -600,32 +608,38 @@ defmodule Phoenix.LiveView.HTMLEngine do
     validate_phx_attrs!(attrs, tag_meta, state)
     attrs = remove_phx_no_break(attrs)
 
-    case pop_special_attr!(attrs, ":for", state) do
-      {{":for", expr, _meta}, attrs} ->
-        state
-        |> update_subengine(:handle_begin, [])
-        |> set_root_on_not_tag()
-        |> push_tag({:tag_open, name, attrs, Map.put(tag_meta, :for, {expr, state})})
-        |> handle_tag_and_attrs(name, attrs, ">", to_location(tag_meta))
+    {new_meta, attrs} =
+      Enum.reduce([:for, :if], {tag_meta, attrs}, fn attr, {meta_acc, attrs_acc} ->
+        string_attr = ":#{attr}"
 
-      nil ->
-        state
-        |> set_root_on_tag()
-        |> push_tag(token)
-        |> handle_tag_and_attrs(name, attrs, ">", to_location(tag_meta))
+        case pop_special_attr!(attrs_acc, string_attr, state) do
+          {{^string_attr, expr, _meta}, attrs} -> {Map.put(meta_acc, attr, {expr, state}), attrs}
+          nil -> {meta_acc, attrs_acc}
+        end
+      end)
+
+    if new_meta == tag_meta do
+      state
+      |> set_root_on_tag()
+      |> push_tag(token)
+      |> handle_tag_and_attrs(name, attrs, ">", to_location(tag_meta))
+    else
+      state
+      |> update_subengine(:handle_begin, [])
+      |> set_root_on_not_tag()
+      |> push_tag({:tag_open, name, attrs, new_meta})
+      |> handle_tag_and_attrs(name, attrs, ">", to_location(new_meta))
     end
   end
 
   defp handle_token({:tag_close, name, tag_meta} = token, state) do
     {{:tag_open, _name, _attrs, tag_open_meta}, state} = pop_tag!(state, token)
     state = update_subengine(state, :handle_text, [to_location(tag_meta), "</#{name}>"])
 
-    case tag_open_meta[:for] do
-      {expr, outer_state} ->
-        handle_for_expr(state, expr, outer_state)
-
-      nil ->
-        state
+    if tag_open_meta[:for] || tag_open_meta[:if] do
+      handle_special_expr(state, tag_open_meta)
+    else
+      state
     end
```

Comment:
Might as well remove this conditional here and add a catch-all clause to `handle_special_expr`?

```suggestion
    handle_special_expr(state, tag_open_meta)
```